posts/ebpf-and-kernel-modules-whats-the-difference.md

```diff
@@ -0,0 +1,115 @@
+
+In eBPF's case, eBPF development packages are needed for compilation. Once the eBPF program is compiled, it will most likely be loaded into the eBPF kernel subsystem with the help of a user-space application that loads and manages the program. The user-space code will also likely be distributed with the eBPF code.
+
+In Kernel Module's case, kernel module development packages are needed for compilation. The build system is typically written in a Makefile. Once the binary is compiled, it will need to be loaded with kernel module management command `insmod`.
+
+
+# Supported Programming Languages
+eBPF programs can be written in any language that has a compiler that can output eBPF bytecode. Currently, and as far as I know, those languages are C, Rust, [BPFTrace](https://github.com/iovisor/bpftrace) and [P4](https://github.com/p4lang/p4c/tree/main).
+
+Kernel modules can be written in C or Rust.
+
+As mentioned, eBPF programs are usually launched with user-space applications. These applications can be written in C, Rust, Go, Python, or any language that has libraries that can load eBPF binaries into the eBPF linux subsystem.
+
+# Development Rigidity
+Kernel Modules are very general purpose and can be used to write any type of program. There is no hand-holding or guardrails either. 
+
+eBPF on the other hand is very rigid and structured. There are even different specified eBPF program types. Each program type will then be restricted on the types of data or functions they are allowed to access since each program is geared towards a narrow domain. To get an eBPF program to compile, it must also pass a verifier check for program correctness and complexity.
+
+# Program Bugs and Safety
+- Kernel modules
+	- Some bugs may have the chance of crashing the entire system. For example, [the infamous NULL pointer dereference](https://www.suse.com/support/kb/doc/?id=000021042)
+	- [Lack of tighter security boundaries can risk corrupting the kernel or be more vulnerable to exploits](https://ebpf.io/what-is-ebpf/#ebpfs-impact-on-the-linux-kernel)
+
+- eBPF
+	- The eBPF verifier makes eBPF programming more rigid, but in doing so attempts to detect improper memory access, infinite loops, code-complexity issues such as too large of a program stack (512 byte limit), and everything in between.
+	- Note: Although one goal of eBPF is to be designed with security in mind and to avoid pitfalls of oversight that can occur in kernel module development, security issues have happened ([CVE-2017-16995](https://bugzilla.redhat.com/show_bug.cgi?id=CVE-2017-16995),[CVE-2020-8835] (https://www.zerodayinitiative.com/blog/2020/4/8/cve-2020-8835-linux-kernel-privilege-escalation-via-improper-ebpf-program-verification)) and will likely happen in the future
+
+If you don't know what you're doing when writing a kernel module, you can completely crash your system. However, you _need to know what you're doing_ to get an eBPF program to do any damage.
```

Comment:
Okay I see, more info about above program checker.  I absolutely love this line right here.  Maybe consider moving or copying this line to the intro, it's that good. And it gives the reader motivation to want to read your article and learn more about eBPF and the differences from kernel modules.

Reply:
added the line to the intro - lmk if it flows well